data/blog/2025/2025-03-18-mezzio101-what-defines-a-middleware-architecture.md

```diff
@@ -0,0 +1,123 @@
+
+- Do you need it to pre-process your request by validating the user session or the form input?
+- Do you need to cache or translate the request?
+- Do you need to contact a third party site to enhance the request?
+
+Your custom middleware can handle **any specific logic** your application may require.
+Some useful things to implement are error handling, CORS, routing, translating.
+
+Ultimately, you can split your application into multiple middlewares that process the request before it reaches your main code.
+An added benefit is that the middlewares can be **reused more easily** in other applications.
+Even better, the middlewares **don't need to be written in the same programming language** or coding architecture.
+
+The alternative to the middleware architecture is the MVC architecture which has grown considerably since its inception, is feature rich, modular and reusable.
+Instead of using flexible layers like a middleware, MVC revolves around controllers and actions.
+Unfortunately, MVC's complexity has brought about its downfall, though it is still being used.
+The important bit and the central difference to middleware is that in a MVC, a request targets a single controller.
+Nowadays, the aim is more toward lightweight, modular and standardized frameworks for producing microframeworks and API based solutions which are more suited to a middleware approach.
+
+### How do you set up a middleware pipeline?
+
+The **middleware pipeline** defines the sequence in which the middlewares are executed.
+Setting up the pipeline needs to be done with care to avoid unwanted behavior.
+In some cases, the execution of one middleware can affect another middleware because of the way the request is passed along.
+The main things to consider when setting up the pipeline are:
+
+- Security
+- Performance
+- Functionality
+
+For example, it makes sense to add an error handler early in the pipeline, even as the first entry.
+If your application requires the user to be logged in, it doesn't make sense to process the request before the authentication middleware gives the go-ahead.
+Also, it doesn't make sense to have an authorization middleware placed before an authentication middleware in the pipeline.
+
+Based on the future needs of your application, the **pipeline can be easily updated** to include additional middleware.
+This aspect goes a long way toward making the flow of your application easier to understand and debug.
+After all, debugging and maintaining smaller sections of code is easier than working with a monolith.
+
+### When and how is middleware executed?
+
+Like the name suggests, **middlewares are executed between the request and the server response** handled by your main codebase.
+The classic way of handling a request meant running through a sequence of code that is less adaptive.
+That is, unless you add a bunch of `if-else` statements for all the cases where the execution needs to exit prematurely.
```

Comment:
This is... not really how most MVC frameworks work anymore. They tend to have lifecycle events that allow you to tie into different points of request handling, allowing early exit. It's not really if/else.

Reply:
my inital goal for this article was not really to compare middleware with MVC, but i guess the two are connected now. 

this particular paragraph compares middleware with the classic PHP4 architecture. in reality, it was intended to mock the old way of doing things, before MVC and middleware. so maybe the best solution is to remove the sentence with 'if-else' altogether?

Reply:
Yes, please, @bidi47